File: control_server/service_manager.py
#!/usr/bin/python
from service.message_define import *
from service_status import *
from whisper_service import *
import collections
import datetime
import io
import logging
import os.path
import threading
import random

class ServiceManager(object):
    """
    store domain node register info/status
    not thread safe
    """
    def __init__(self, logger_name):
        self.logger = logging.getLogger(logger_name)
        # #type group list
        # #key = server type, value = {group:list of service name}
        self.server_groups = {}

        # #key = service name, value = service
        self.service_map = {}
        # #key = server uuid, value = [service name]
        self.service_in_server = {}
        # #key = service name, value = [whisper service]
        self.whispers = {}
        self.lock = threading.RLock()

    def activeService(self, service):
        with self.lock:
            if not self.service_map.has_key(service.name):
                # #new service
                # #add new service to group
                if not self.server_groups.has_key(service.type):
                    # #new service type
                    self.server_groups[service.type] = {"default":[service.name]}
                else:
                    self.server_groups[service.type]["default"].append(service.name)

                self.service_map[service.name] = service

                if not self.service_in_server.has_key(service.server):
                    self.service_in_server[service.server] = [service.name]
                else:
                    if service.name not in self.service_in_server[service.server]:
                        self.service_in_server[service.server].append(service.name)

                self.logger.info("<ServiceManager> add new service, service name '%s'" % (
                    service.name))

            current = self.service_map[service.name]
            current.ip = service.ip
            current.port = service.port
            current.version = service.version
            current.status = ServiceStatus.status_running
            current.server = service.server
    # #        self.logger.info("<ServiceManager> service actived, service name '%s'"%(
    # #            current.name))
            return True

    def deactiveService(self, name):
        with self.lock:

            if not self.service_map.has_key(name):
                self.logger.error("<ServiceManager> deactive service fail, invalid service '%s'" % (name))
                return False

            if self.containsWhisper(name):
                self.removeWhisper(name)

            service = self.service_map[name]
            service.status = ServiceStatus.status_stop
            self.logger.info("<ServiceManager> service deactived, service name '%s'" % (name))
            return True

# #    def modifyService(self, name, service):
# #        pass

    def containsService(self, service_name):
        with self.lock:
            return self.service_map.has_key(service_name)

    def getService(self, service_name):
        with self.lock:
            if not self.service_map.has_key(service_name):
                return None
            else:
                return self.service_map[service_name]

    def loadService(self, service_type, service_list):
        with self.lock:

            if self.server_groups.has_key(service_type):
                new_service_map = {}
                for service in service_list:
                    new_service_map[service.name] = service
                    
                # #clear all
                for name_list in self.server_groups[service_type].values():
                    for service_name in name_list:
                        # remove service name in service_in_server
                        if self.service_map.has_key(service_name):
                            server = self.service_map[service_name].server
                            if self.service_in_server.has_key(server):
                                if service_name in self.service_in_server[server]:
                                    self.service_in_server[server].remove(service_name)
                                    if len(self.service_in_server[server]) == 0:
                                        del self.service_in_server[server]
                        
                        ## copy disk type
                        if new_service_map.has_key(service_name):
                            new_service = new_service_map[service_name]
                            disk_type = self.service_map[service_name].disk_type
                            new_service.disk_type = disk_type
                        # remove service name in service_map
                        del self.service_map[service_name]

            self.server_groups[service_type] = {"default":[]}

            for service in service_list:
                ## copy disk type
                if self.service_map.has_key(service.name):
                    old_service = self.service_map[service.name]
                    service.disk_type = old_service.disk_type
                
                self.service_map[service.name] = service
                self.server_groups[service_type]["default"].append(service.name)

                if not self.service_in_server.has_key(service.server):
                    self.service_in_server[service.server] = [service.name]
                else:
                    if service.name not in self.service_in_server[service.server]:
                        self.service_in_server[service.server].append(service.name)

            self.logger.info("<ServiceManager> %d service(s) loaded for service type %d" % (
                len(service_list), service_type))
            return True

    def queryService(self, service_type, service_group):
        with self.lock:
            result = []
            if not self.server_groups.has_key(service_type):
                return result
            group_dict = self.server_groups[service_type]
            if not group_dict.has_key(service_group):
                return result
            for name in group_dict[service_group]:
                if not self.service_map.has_key(name):
                    continue

                result.append(self.service_map[name])

            return result
        
    def updateService(self, name, disk_type):
        with self.lock:
            if self.service_map.has_key(name):
                self.service_map[name].disk_type = disk_type
            else:
                service = ServiceStatus()
                service.name = name
                service.disk_type = disk_type
                self.service_map[service.name] = service

    def queryServicesInServer(self, server_id):
        with self.lock:
            result = []
            if self.service_in_server.has_key(server_id):
                for service_name in self.service_in_server[server_id]:
                    result.append(service_name)

            return result

    def getAllServiceType(self):
        with self.lock:
            return self.server_groups.keys()

    def queryServiceGroup(self, service_type):
        with self.lock:
            result = []
            if self.server_groups.has_key(service_type):
                result.extend(self.server_groups[service_type].keys())

            return result

    def updateWhisper(self, service_name, whisper_list):
        with self.lock:
            if not self.service_map.has_key(service_name):
                return False
            # # remove all
            self.whispers[service_name] = []
            service = self.service_map[service_name]
            for whisper in whisper_list:
                whisper.name = service_name
                whisper.type = service.type
                self.whispers[service_name].append(whisper)

            return True

    def removeWhisper(self, service_name):
        with self.lock:
            if self.whispers.has_key(service_name):
                del self.whispers[service_name]

    def getAllWhisper(self):
        with self.lock:
            result = []
            for whisper_list in self.whispers.values():
                result.extend(whisper_list)

            return result

    def containsWhisper(self, service_name):
        with self.lock:
            return (self.whispers.has_key(service_name))

    def getWhisper(self, service_name):
        with self.lock:
            if self.whispers.has_key(service_name):
                # #load balance
                whisper_list = self.whispers[service_name]
                index = random.randint(0, len(whisper_list) - 1)
                whisper = whisper_list[index]

                return whisper
            else:
                return None

    def statisticStatus(self):
        with self.lock:
            stop = 0
            warn = 0
            error = 0
            run = 0
            for service in self.service_map.values():
                if service.status == ServiceStatus.status_stop:
                    stop += 1
                elif service.status == ServiceStatus.status_warning:
                    warn += 1
                elif service.status == ServiceStatus.status_error:
                    error += 1
                else:
                    run += 1

            return [stop, warn, error, run]

File: control_server_cpp/src/service_manager.cpp
#include "service_manager.h"


using namespace zhicloud;
using namespace control_server;
using namespace util;

ServiceManager::ServiceManager(const string& logger_name)
{
   _logger = getLogger(logger_name);
}

ServiceManager::~ServiceManager()
{
    //dtor
}

void ServiceManager::addService(const ServiceStatus& service)
{
    if(_server_groups.count(service.getType()) == 0){

        map<string,set<string>> tmp;
        tmp["default"] = move(set<string>());
        _server_groups[service.getType()] = move(tmp);

    }
    _server_groups[service.getType()]["default"].emplace(service.getName());

    if(_service_in_server.count(service.getServer()) == 0){
        _service_in_server[service.getServer()] = move(set<string>());
    }

    _service_in_server[service.getServer()].emplace(service.getName());
    _service_map.emplace(service.getName(),service);//add new service
}

bool ServiceManager::activeService(const ServiceStatus& service)
{
    lock_type lock(_mutex);
    if(_service_map.count(service.getName()) == 0)
    {
        addService(service);
    }

    auto item = _service_map.find(service.getName());
    if( item == _service_map.end()){
        return false;
    }

    item->second.setIp(service.getIp());
    item->second.setPort(service.getPort());
    item->second.setVersion(service.getVersion());
    item->second.setServer(service.getServer());
    item->second.setStatus(UnitStatusEnum::status_running);


    return true;
}

bool ServiceManager::deactiveService(const ServiceStatus& service)
{
    lock_type lock(_mutex);
    auto item = _service_map.find(service.getName());
    if( item == _service_map.end()){
        return false;
    }

    auto iter = _whispers.find(service.getName());
    if(iter != _whispers.end())
        _whispers.erase(iter);

    item->second.setStatus(UnitStatusEnum::status_stop);
    return true;
}

bool ServiceManager::containsService(const string& name)
{
    lock_type lock(_mutex);
    return (_service_map.count(name) == 1 ? true : false);
}

bool ServiceManager::getService(const string& name,ServiceStatus& service)const
{
    lock_type lock(_mutex);
    auto item = _service_map.find(service.getName());
    if( item != _service_map.end()){
        service = item->second;
        return true;
    }
    return false;
}

void ServiceManager::deleteServiceFromServer(const UUID_TYPE& server_id,const string& service_name)
{
    auto service_list = _service_in_server.find(server_id);
    if(service_list != _service_in_server.end()){

        auto del_service = service_list->second.find(service_name);
        if(del_service != service_list->second.end())
            service_list->second.erase(del_service);

        if(service_list->second.size() == 0)
            _service_in_server.erase(service_list);

    }


}

void ServiceManager::loadService(const uint32_t& service_type, const vector<ServiceStatus>& service_list)
{
    lock_type lock(_mutex);
    map<string,ServiceStatus> new_service_map;
    for(const auto& service : service_list)
        new_service_map.emplace(service.getName(),service);

    //copy disk type
    for(auto& service : new_service_map)
    {
        auto old_service =_service_map.find(service.second.getName());
        if(old_service !=_service_map.end())
        {
            service.second.setDiskType(old_service->second.getDiskType());

        }
    }

    if(_server_groups.count(service_type))
    {
        //begin clear
        for(const auto &group : _server_groups[service_type])//map<group,set<service>>
        {
            for(const auto & service_name : group.second) //set<service>
            {
                auto old_service = _service_map.find(service_name);
                if(old_service != _service_map.end())
                {
                    deleteServiceFromServer(old_service->second.getServer(),service_name);
                    _service_map.erase(old_service);
                }

            }
        }

         auto default_group = _server_groups[service_type].find("default");
         if(default_group  != _server_groups[service_type].end())
            default_group->second.clear();

    }

    for(auto& service : new_service_map)
    {
        auto old_service =_service_map.find(service.first);
        if( old_service !=_service_map.end())
            _service_map.erase(old_service);

        addService(service.second);
    }

}

void ServiceManager::queryService(const uint32_t& service_type,const string& group_name,vector<ServiceStatus>& status_list)const
{
    lock_type lock(_mutex);
    //map<service_type,map<group_name,set<service_name>>> _server_groups;

    auto group =_server_groups.find(service_type);
    if( group == _server_groups.end())
        return;

    auto service_list = group->second.find(group_name);
    if( service_list == group->second.end())
        return;


    for(const auto& name : service_list->second)
    {
        const auto status = _service_map.find(name);
        if(status != _service_map.end()){
            status_list.emplace_back(status->second);
        }

    }
}

bool ServiceManager::updateService(const string& service_name,const ComputeStorageTypeEnum& type)
{
    lock_type lock(_mutex);
    auto item = _service_map.find(service_name);
    if( item != _service_map.end()){
        item->second.setDiskType(type);
        return true;
    }
    return false;
}


bool ServiceManager::queryServicesInServer(const UUID_TYPE& server,vector<string>& service_list)const
{
    lock_type lock(_mutex);
    auto item = _service_in_server.find(server);
    if( item != _service_in_server.end()){

        for(const auto & name :item->second)
            service_list.emplace_back(name);

        return true;
    }
    return false;
}

void ServiceManager::getAllServiceType(vector<uint32_t>& type_list)const
{
    lock_type lock(_mutex);
    for(const auto& item : _server_groups)
    {
        type_list.emplace_back(item.first);
    }
}

bool ServiceManager::queryServiceGroup(const uint32_t& service_type,vector<string>& group_list)const
{
    lock_type lock(_mutex);
    auto item = _server_groups.find(service_type);
    if( item != _server_groups.end()){

        for(const auto& group : item->second)
            group_list.emplace_back(group.first);

        return true;
    }
    return false;
}

 bool ServiceManager::updateWhisper(const string& service_name,const vector<WhisperService>& whisper_list)
 {
    lock_type lock(_mutex);
    auto item = _service_map.find(service_name);
    if( item != _service_map.end()){

        auto whisper_item =_whispers.find(service_name);
        if(whisper_item == _whispers.end())
        {
            _whispers[service_name] = move(vector<WhisperService>());
        }else{
            whisper_item->second.clear();
        }

        for(const auto& whisper : whisper_list){
            WhisperService tmp(whisper);
            tmp.setName(service_name);
            tmp.setType(item->second.getType());
            _whispers[service_name].emplace_back(tmp);
        }


        return true;
    }
    return false;
 }
bool ServiceManager::containsWhisper(const string& service_name)
{
    lock_type lock(_mutex);
    return (_whispers.count(service_name) == 1 ? true: false);
}

bool ServiceManager::getWhisper(const string& service_name,vector<WhisperService>& whisper_list)const
{
    lock_type lock(_mutex);
    auto item = _whispers.find(service_name);
    if( item != _whispers.end()){
        for(const auto &whisper : item->second)
        {
            whisper_list.emplace_back(whisper);
        }
        return true;
    }
    return false;
}

void ServiceManager::getAllWhisper(vector<WhisperService>& whisper_list)const
{
    lock_type lock(_mutex);
    for(const auto &item : _whispers)
    {
         for(const auto &whisper : item.second)
            whisper_list.emplace_back(whisper);
    }
}

void ServiceManager::statisticStatus(StatusCounter &counter)const
{
    lock_type lock(_mutex);
    for(const auto &item : _service_map)
    {
        switch(item.second.getStatus())
        {
            case UnitStatusEnum::status_stop:
            counter.IncreaseStop();
            break;
            case UnitStatusEnum::status_running:
            counter.IncreaseRun();
            break;
            case UnitStatusEnum::status_warning:
            counter.IncreaseWarn();
            break;
            case UnitStatusEnum::status_error:
            counter.IncreaseError();
            break;
            default:
            break;
        }
    }
}
